### PR TITLE
Validate annotations

### PR DESCRIPTION
**Pull request by oaons**
Trying to solve https://github.com/ahrefs/atd/issues/227 and https://github.com/ahrefs/atd/issues/204

- [x] added straightforward validation of `json` field names + tests. Now invalid field name will throw an error on `atdgen -j ...`
- [x] removed tests related to retired `json.tag_field` field since now they generate errors

### CODE REVIEW

**Comment by rr0gi** [comment]
I have a gut feeling it should be a warning, not a hard error. E.g. : atd can have annotations for different languages, atdgen may not know all annotations interpreted by other generators.

File: atdgen/src/json.ml [hunk 1]
```diff
@@ -63,6 +63,26 @@ type json_repr =
...
+  "name";
+  "keep_nulls"
+]
+
+let validate_json_annotation an =
+  List.iter (fun (s, (_, fs)) ->
+    if s = "json" then
+      List.iter (fun (f, (loc, _)) ->
+        if (not (List.mem f json_fields)) then
+          Error.error loc (Printf.sprintf "json.%s annotation not allowed" f)
```
Comment (rr0gi) [inline atdgen/src/json.ml:82]: s/not allowed/is not recognized/

**Comment by rr0gi** [comment]
though this argument doesn't apply to json annotations.. Another weak argument - this removed tag_field annotation, we may want to make it possible process old atds without causing build breakage (it will still break later if code expects tag_field to work)

**Comment by mjambon** [comment]
Hey all, thanks for starting this. I implemented schema-based validation for easily specifying which annotations of a given kind are allowed where. The PR is here: #244

**Comment by cyberhuman** [comment]
Superseded by #244.

**Pull request closed without merging**